be/src/connector/binlog_connector.cpp

```diff
@@ -0,0 +1,212 @@
+    fields.emplace(BINLOG_SEQ_ID, seq_id_field);
+    cid += 1;
+
+    VectorizedFieldPtr timestamp_field = std::make_shared<VectorizedField>(
+            cid, BINLOG_TIMESTAMP, get_type_info(TYPE_BIGINT), STORAGE_AGGREGATE_NONE, 8, false, false);
+    fields.emplace(BINLOG_TIMESTAMP, timestamp_field);
+    cid += 1;
+
+    return fields;
+}
+
+StatusOr<VectorizedSchema> BinlogDataSource::_build_binlog_schema() {
+    BinlogMetaFieldMap binlog_meta_map = _build_binlog_meta_fields(_tablet->tablet_schema().num_columns());
+    std::vector<uint32_t> data_column_cids;
+    std::vector<uint32_t> meta_column_slot_index;
+    VectorizedFields meta_fields;
+    int slot_index = -1;
+    for (auto slot : _tuple_desc->slots()) {
+        DCHECK(slot->is_materialized());
+        slot_index += 1;
+        int32_t index = _tablet->field_index(slot->col_name());
+        if (index >= 0) {
+            data_column_cids.push_back(index);
+        } else if (slot->col_name() == BINLOG_OP) {
+            meta_column_slot_index.push_back(slot_index);
+            meta_fields.emplace_back(binlog_meta_map[BINLOG_OP]);
+        } else if (slot->col_name() == BINLOG_VERSION) {
+            meta_column_slot_index.push_back(slot_index);
+            meta_fields.emplace_back(binlog_meta_map[BINLOG_VERSION]);
+        } else if (slot->col_name() == BINLOG_SEQ_ID) {
+            meta_column_slot_index.push_back(slot_index);
+            meta_fields.emplace_back(binlog_meta_map[BINLOG_SEQ_ID]);
+        } else if (slot->col_name() == BINLOG_TIMESTAMP) {
+            meta_column_slot_index.push_back(slot_index);
+            meta_fields.emplace_back(binlog_meta_map[BINLOG_TIMESTAMP]);
+        } else {
+            std::stringstream ss;
+            ss << "invalid field name: " << slot->col_name();
+            LOG(WARNING) << ss.str();
+            return Status::InternalError(ss.str());
+        }
+    }
+
+    if (data_column_cids.empty()) {
+        return Status::InternalError("failed to build binlog schema, no materialized data slot!");
+    }
+
+    const TabletSchema& tablet_schema = _tablet->tablet_schema();
+    VectorizedSchema schema = ChunkHelper::convert_schema_to_format_v2(tablet_schema, data_column_cids);
+    for (int32_t i = 0; i < meta_column_slot_index.size(); i++) {
+        uint32_t index = meta_column_slot_index[i];
+        if (index >= schema.num_fields()) {
+            schema.append(meta_fields[i]);
+        } else {
+            schema.insert(index, meta_fields[i]);
+        }
+    }
+
+    return schema;
+}
+
+Status BinlogDataSource::_mock_chunk(Chunk* chunk) {
+    int32_t num = _chunk_num.fetch_add(1, std::memory_order_relaxed);
+    if (num >= 2) {
+        return Status::EndOfFile(fmt::format("Has sent {} chunks", num));
+    }
+
+    for (int row = 0; row < 10; row++) {
+        for (int col = 0; col < chunk->num_columns(); col++) {
+            LogicalType type = _binlog_read_schema.field(col)->type()->type();
```

Comment:
we can not 'select a from xxx _BINGLOG' if a is a data column which is varchar type, am i right?

Reply:
Yes, you are right. This PR only mock data for some types, and it's enough for testing. Of course we can mock data for more types if needed.